rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,929 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Python.h>
+
+#include <rcl/error_handling.h>
+#include <rcl_action/rcl_action.h>
+
+#include "./impl/common.h"
+
+/// Destroy an rcl_action entity.
+/**
+ * Raises RuntimeError on failure.
+ *
+ * \param[in] pyentity Capsule pointing to the entity to destroy.
+ * \param[in] pynode Capsule pointing to the node the action client was added to.
+ */
+static PyObject *
+rclpy_action_destroy_entity(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyentity;
+  PyObject * pynode;
+
+  if (!PyArg_ParseTuple(args, "OO", &pyentity, &pynode)) {
+    return NULL;
+  }
+
+  rcl_node_t * node = (rcl_node_t *)PyCapsule_GetPointer(pynode, "rcl_node_t");
+  if (!node) {
+    return NULL;
+  }
+
+  rcl_ret_t ret;
+  if (PyCapsule_IsValid(pyentity, "rcl_action_client_t")) {
+    rcl_action_client_t * action_client =
+      (rcl_action_client_t *)PyCapsule_GetPointer(pyentity, "rcl_action_client_t");
+    ret = rcl_action_client_fini(action_client, node);
+    PyMem_Free(action_client);
+  } else if (PyCapsule_IsValid(pyentity, "rcl_action_server_t")) {
+    rcl_action_server_t * action_server =
+      (rcl_action_server_t *)PyCapsule_GetPointer(pyentity, "rcl_action_server_t");
+    ret = rcl_action_server_fini(action_server, node);
+    PyMem_Free(action_server);
+  } else {
+    ret = RCL_RET_ERROR;  // to avoid a linter warning
+    const char * entity_name = PyCapsule_GetName(pyentity);
+    if (!entity_name) {
+      return NULL;
+    }
+    return PyErr_Format(PyExc_RuntimeError, "'%s' is not a known entity", entity_name);
+  }
+
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to fini '%s': %s", PyCapsule_GetName(pyentity), rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  if (PyCapsule_SetPointer(pyentity, Py_None)) {
+    // exception set by PyCapsule_SetPointer
+    return NULL;
+  }
+
+  Py_RETURN_NONE;
+}
+
+/// Fetch a predefined qos_profile from rcl_action and convert it to a Python QoSProfile object.
+/**
+ * Raises RuntimeError if the QoS profile is unknown.
+ *
+ * This function takes a string defining a rmw_qos_profile_t and returns the
+ * corresponding Python QoSProfile object.
+ * \param[in] pyrmw_profile String with the name of the profile to load.
+ * \return QoSProfile object.
+ */
+static PyObject *
+rclpy_action_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  const char * pyrmw_profile;
```

Comment:
I would remove the `py` prefix since it's not a `PyObject`. There isn't a rule for this in the style guide AFAIK though

Reply:
Done 34db026